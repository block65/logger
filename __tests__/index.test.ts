import {
  afterAll,
  beforeEach,
  describe,
  expect,
  jest,
  test,
} from '@jest/globals';
import { createLoggerWithWaitableMock } from './helpers.js';

describe('Basic', () => {
  const oldEnv = process.env;
  const originalStackTraceLimit = Error.stackTraceLimit;

  beforeEach(() => {
    jest.resetModules();
    jest.useFakeTimers('modern');
    jest.setSystemTime(new Date('2009-02-13T23:31:30.000Z'));
    process.env = { ...oldEnv };
    Error.stackTraceLimit = originalStackTraceLimit;
  });

  afterAll(() => {
    process.env = oldEnv;
    Error.stackTraceLimit = originalStackTraceLimit;
  });

  test('Object', async () => {
    const [logger, callback] = await createLoggerWithWaitableMock();
    logger.warn({ omg: true });
    await expect(callback.waitUntilCalled()).resolves.toMatchSnapshot();
  });

  test('Object with undefined props', async () => {
    const [logger, callback] = await createLoggerWithWaitableMock();
    logger.warn({ omg: undefined });
    await expect(callback.waitUntilCalled()).resolves.toMatchSnapshot();
  });

  test('String, Object', async () => {
    const [logger, callback] = await createLoggerWithWaitableMock();
    logger.warn('hello', { omg: true });
    await expect(callback.waitUntilCalled()).resolves.toMatchSnapshot();
  });

  test('String Format, Object', async () => {
    const [logger, callback] = await createLoggerWithWaitableMock();
    logger.warn('hello %o', { omg: true });
    await expect(callback.waitUntilCalled()).resolves.toMatchSnapshot();
  });

  test('Object, String Format, String', async () => {
    const [logger, callback] = await createLoggerWithWaitableMock();
    logger.warn({ omg: true }, 'hello %s', 'world');
    await expect(callback.waitUntilCalled()).resolves.toMatchSnapshot();
  });

  test('Object, String Format, String', async () => {
    const [logger, callback] = await createLoggerWithWaitableMock();

    logger.warn({ omg: true }, 'hello %s:%s %d', 'world', 'and', 123);
    await expect(callback.waitUntilCalled()).resolves.toMatchSnapshot();
  });

  test('Object + String = Format', async () => {
    const [logger, callback] = await createLoggerWithWaitableMock();
    logger.warn({ omg: true }, 'hello');
    await expect(callback.waitUntilCalled()).resolves.toMatchSnapshot();
  });

  test('Trace Caller Auto (development)', async () => {
    process.env.NODE_ENV = 'development';
    const [logger, callback] = await createLoggerWithWaitableMock();
    logger.warn('hello');
    await expect(callback.waitUntilCalled()).resolves.toMatchSnapshot();
  });

  test('Trace Caller Auto (production)', async () => {
    process.env.NODE_ENV = 'production';
    Error.stackTraceLimit = 10;
    const [logger, callback] = await createLoggerWithWaitableMock();
    logger.warn('hello');
    await expect(callback.waitUntilCalled()).resolves.toMatchSnapshot();
  });

  test('Trace Caller Force true (production)', async () => {
    process.env.NODE_ENV = 'production';
    Error.stackTraceLimit = 10;
    const [logger, callback] = await createLoggerWithWaitableMock({
      traceCaller: true,
    });
    logger.warn('hello');
    await expect(callback.waitUntilCalled()).resolves.toMatchSnapshot();
  });

  test('Trace Caller Force false (development)', async () => {
    process.env.NODE_ENV = 'development';
    Error.stackTraceLimit = 10;
    const [logger, callback] = await createLoggerWithWaitableMock({
      traceCaller: false,
    });
    logger.warn('hello');
    await expect(callback.waitUntilCalled()).resolves.toMatchSnapshot();
  });

  test('Mixins basic + accum', async () => {
    const [logger, callback] = await createLoggerWithWaitableMock({
      mixins: [
        () => ({
          logger: {
            go: 'log',
          },
        }),
        () => ({
          luger: {
            go: 'bang',
          },
        }),
        (accum) => ({
          accum,
        }),
      ],
    });
    logger.warn('hello');
    await expect(callback.waitUntilCalled()).resolves.toMatchSnapshot();
  });

  test('Mixins property overwrite', async () => {
    const [logger, callback] = await createLoggerWithWaitableMock({
      mixins: [
        () => ({
          logger: {
            go: 'log',
          },
        }),
        () => ({
          logger: {
            go: 'brrr',
          },
        }),
      ],
    });
    logger.warn('hello');
    await expect(callback.waitUntilCalled()).resolves.toMatchSnapshot();
  });

  test('allow undefined logger level', async () => {
    await createLoggerWithWaitableMock({
      level: undefined,
    });
  });
});
