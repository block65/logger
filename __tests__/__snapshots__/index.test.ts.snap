// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic Error Object 1`] = `
Object {
  "err": Object {
    "debug": "wooyeah",
    "message": "hallo",
    "stack": "Error: hallo
    at Object.<anonymous> (/srv/test/__tests__/index.test.ts:12:32)
    at Promise.then.completed (/srv/test/node_modules/jest-circus/build/utils.js:390:28)
    at new Promise (<anonymous>)
    at callAsyncCircusFn (/srv/test/node_modules/jest-circus/build/utils.js:315:10)
    at _callCircusTest (/srv/test/node_modules/jest-circus/build/run.js:218:40)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at _runTest (/srv/test/node_modules/jest-circus/build/run.js:155:3)
    at _runTestsForDescribeBlock (/srv/test/node_modules/jest-circus/build/run.js:66:9)
    at _runTestsForDescribeBlock (/srv/test/node_modules/jest-circus/build/run.js:60:9)
    at run (/srv/test/node_modules/jest-circus/build/run.js:25:3)
    at runAndTransformResultsToJestFormat (/srv/test/node_modules/jest-circus/build/legacy-code-todo-rewrite/jestAdapterInit.js:166:21)
    at jestAdapter (/srv/test/node_modules/jest-circus/build/legacy-code-todo-rewrite/jestAdapter.js:82:19)
    at runTestInternal (/srv/test/node_modules/jest-runner/build/runTest.js:389:16)
    at runTest (/srv/test/node_modules/jest-runner/build/runTest.js:481:34)",
    "type": "Error",
  },
  "hostname": "testhost",
  "level": 50,
  "msg": "hallo",
  "pid": 1337,
  "time": 1234567890000,
}
`;

exports[`Basic Error Object serialized on non-error level 1`] = `
Object {
  "err": Object {
    "message": "hallo",
    "stack": "Error: hallo
    at Object.<anonymous> (/srv/test/__tests__/index.test.ts:18:17)
    at Promise.then.completed (/srv/test/node_modules/jest-circus/build/utils.js:390:28)
    at new Promise (<anonymous>)
    at callAsyncCircusFn (/srv/test/node_modules/jest-circus/build/utils.js:315:10)
    at _callCircusTest (/srv/test/node_modules/jest-circus/build/run.js:218:40)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at _runTest (/srv/test/node_modules/jest-circus/build/run.js:155:3)
    at _runTestsForDescribeBlock (/srv/test/node_modules/jest-circus/build/run.js:66:9)
    at _runTestsForDescribeBlock (/srv/test/node_modules/jest-circus/build/run.js:60:9)
    at run (/srv/test/node_modules/jest-circus/build/run.js:25:3)
    at runAndTransformResultsToJestFormat (/srv/test/node_modules/jest-circus/build/legacy-code-todo-rewrite/jestAdapterInit.js:166:21)
    at jestAdapter (/srv/test/node_modules/jest-circus/build/legacy-code-todo-rewrite/jestAdapter.js:82:19)
    at runTestInternal (/srv/test/node_modules/jest-runner/build/runTest.js:389:16)
    at runTest (/srv/test/node_modules/jest-runner/build/runTest.js:481:34)",
    "type": "Error",
  },
  "hostname": "testhost",
  "level": 30,
  "msg": "hallo",
  "pid": 1337,
  "time": 1234567890000,
}
`;

exports[`Basic Mixins basic + accum 1`] = `
Object {
  "accum": Object {
    "logger": Object {
      "go": "log",
    },
    "luger": Object {
      "go": "bang",
    },
  },
  "caller": "Object.<anonymous> (__tests__/index.test.ts:108:12)",
  "hostname": "testhost",
  "level": 40,
  "logger": Object {
    "go": "log",
  },
  "luger": Object {
    "go": "bang",
  },
  "msg": "hello",
  "pid": 1337,
  "time": 1234567890000,
}
`;

exports[`Basic Mixins property overwrite 1`] = `
Object {
  "caller": "Object.<anonymous> (__tests__/index.test.ts:127:12)",
  "hostname": "testhost",
  "level": 40,
  "logger": Object {
    "go": "brrr",
  },
  "msg": "hello",
  "pid": 1337,
  "time": 1234567890000,
}
`;

exports[`Basic Object + String = Format 1`] = `
Object {
  "hostname": "testhost",
  "level": 40,
  "msg": "hello",
  "omg": true,
  "pid": 1337,
  "time": 1234567890000,
}
`;

exports[`Basic Object 1`] = `
Object {
  "hostname": "testhost",
  "level": 40,
  "omg": true,
  "pid": 1337,
  "time": 1234567890000,
}
`;

exports[`Basic Object, String Format, String 1`] = `
Object {
  "hostname": "testhost",
  "level": 40,
  "msg": "hello world",
  "omg": true,
  "pid": 1337,
  "time": 1234567890000,
}
`;

exports[`Basic Object, String Format, String 2`] = `
Object {
  "hostname": "testhost",
  "level": 40,
  "msg": "hello world:and 123",
  "omg": true,
  "pid": 1337,
  "time": 1234567890000,
}
`;

exports[`Basic String Format, Object 1`] = `
Object {
  "hostname": "testhost",
  "level": 40,
  "msg": "hello {\\"omg\\":true}",
  "pid": 1337,
  "time": 1234567890000,
}
`;

exports[`Basic String, Object 1`] = `
Object {
  "hostname": "testhost",
  "level": 40,
  "msg": "hello",
  "pid": 1337,
  "time": 1234567890000,
}
`;

exports[`Basic Trace Caller Force (production) 1`] = `
Object {
  "caller": "Object.<anonymous> (__tests__/index.test.ts:77:12)",
  "hostname": "testhost",
  "level": 40,
  "msg": "hello",
  "pid": 1337,
  "time": 1234567890000,
}
`;
